PluginsAndFeatures/azure-toolkit-for-intellij/src/main/java/com/microsoft/azure/toolkit/intellij/function/runner/core/FunctionCliResolver.java

```diff
@@ -5,73 +5,151 @@
 
 package com.microsoft.azure.toolkit.intellij.function.runner.core;
 
-import com.microsoft.azure.toolkit.lib.common.logging.Log;
-import com.microsoft.azure.toolkit.lib.common.utils.TextUtils;
-import com.microsoft.azuretools.utils.CommandUtils;
+import com.microsoft.azure.toolkit.lib.common.cache.Cacheable;
+import com.microsoft.azure.toolkit.lib.common.operation.AzureOperation;
+import com.microsoft.azure.toolkit.lib.common.utils.CommandUtils;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang3.StringUtils;
 
+import javax.annotation.Nullable;
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Optional;
+import java.util.Set;
 
-import static com.microsoft.intellij.ui.messages.AzureBundle.message;
-
-public class FunctionCliResolver {
-    public static String resolveFunc() throws IOException, InterruptedException {
-        final boolean isWindows = CommandUtils.isWindows();
-        final List<File> funCmdFiles = CommandUtils.resolvePathForCommand("func");
-        File result = null;
-        for (final File file : funCmdFiles) {
-            final File canonicalFile = file.getCanonicalFile();
-            if (!canonicalFile.exists()) {
-                continue;
-            }
-            // when `func core tools` is manually installed and func is available at PATH
-            // use canonical path to locate the real installation path
-            result = findFuncExecInFolder(canonicalFile.getParentFile(), isWindows);
-            if (result == null) {
-                if (isWindows) {
-                    result = resolveFuncForWindows(canonicalFile);
-                } else {
-                    // in linux/mac, when the way of `npm install azure-functions-core-tools`, the canonicalFile will point to `main.js`
-                    if (canonicalFile.getName().equals("main.js")) {
-                        result = findFuncExecInFolder(Paths.get(canonicalFile.getParent(), "..", "bin").normalize().toFile(),
-                                isWindows);
-                    }
+import static com.microsoft.azure.toolkit.lib.common.utils.CommandUtils.isWindows;
+
+public abstract class FunctionCliResolver {
+    public static String resolveFunc() {
+        return resolve().stream().findFirst().orElse(null);
+    }
+
+    @AzureOperation(name = "function.resolve_cli", type = AzureOperation.Type.TASK)
+    public static List<String> resolve() {
+        final FunctionCliResolver resolver = CommandUtils.isWindows() ? new WindowsFunctionCoreToolsResolver() : new LinuxFunctionCoreToolsResolver();
```

Comment:
how about to define static windowFunctionCliResolver and linuxFunctionCliResolver to avoid creating them every time?

Reply:
or, I think, to make `resolveFunc` not static, and defining a single resolver should be better.

Reply:
how to inherit methods in static method? I can create a final static  resolver to avoid call ctor every time

Reply:
that's good